tpm2tools/signer.go

```diff
@@ -88,29 +77,92 @@ func (k *Key) GetSigner() (crypto.Signer, error) {
 	if k.hasAttribute(tpm2.FlagRestricted) {
 		return nil, fmt.Errorf("restricted keys are not supported")
 	}
+	hashAlg, err := getSigningHashAlg(k)
+	if err != nil {
+		return nil, err
+	}
+	// For crypto.Signer, Go does the hashing. Make sure the hash is supported.
+	hash, err := hashAlg.Hash()
+	if err != nil {
+		return nil, err
+	}
+	return &tpmSigner{k, hash}, nil
+}
+
+// SignData TODO(joerichey) fill this out
+func (k *Key) SignData(data []byte) ([]byte, error) {
+	hashAlg, err := getSigningHashAlg(k)
+	if err != nil {
+		return nil, err
+	}
+
+	var digest []byte
+	var ticket *tpm2.Ticket
+	if k.hasAttribute(tpm2.FlagRestricted) {
+		// Restricted keys can only sign data hashed by the TPM. We use the
+		// owner hierarchy for the Ticket, but any non-Null hierarchy would do.
+		digest, ticket, err = tpm2.Hash(k.rw, hashAlg, data, tpm2.HandleOwner)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Unrestricted keys can sign arbitrary digests, no need for TPM call.
+		hash, err := hashAlg.Hash()
+		if err != nil {
+			return nil, err
+		}
+		hasher := hash.New()
+		hasher.Write(data)
+		digest = hasher.Sum(nil)
+	}
+
+	auth, err := k.session.Auth()
+	if err != nil {
+		return nil, err
+	}
+	sig, err := tpm2.SignWithSession(k.rw, auth.Session, k.handle, "", digest, ticket, nil)
+	if err != nil {
+		return nil, err
+	}
+	return getSignature(sig)
+}
+
+func getSigningHashAlg(k *Key) (tpm2.Algorithm, error) {
 	if !k.hasAttribute(tpm2.FlagSign) {
-		return nil, fmt.Errorf("non-signing key used with GetSigner()")
+		return tpm2.AlgNull, fmt.Errorf("non-signing key used with signing operation")
 	}
 
 	var sigScheme *tpm2.SigScheme
-
 	switch k.pubArea.Type {
 	case tpm2.AlgRSA:
 		sigScheme = k.pubArea.RSAParameters.Sign
-		if sigScheme.Alg != tpm2.AlgRSAPSS && sigScheme.Alg != tpm2.AlgRSASSA {
-			return nil, fmt.Errorf("unsupported signing algorithm: %v", sigScheme.Alg)
-		}
 	case tpm2.AlgECC:
 		sigScheme = k.pubArea.ECCParameters.Sign
-		if sigScheme.Alg != tpm2.AlgECDSA {
-			return nil, fmt.Errorf("unsupported signing algorithm: %v", sigScheme.Alg)
-		}
 	default:
-		return nil, fmt.Errorf("unsupported key type: %v", k.pubArea.Type)
+		return tpm2.AlgNull, fmt.Errorf("unsupported key type: %v", k.pubArea.Type)
 	}
-	hash, err := sigScheme.Hash.Hash()
-	if err != nil {
-		return nil, err
+
+	if sigScheme == nil {
```

Comment:
OTOH, thinking of "SignData" as the convenient way to sign with attestation keys, maybe this use case can be discarded. WDYT?

Reply:
Do we actually envision a use-case where the same key would be used with different hash algorithms (for example)? I think ideally the hash algorithm would be fixed at key creation time.

If there is a reason to use multiple hash algorithms with the same private signing key, we could solve this problem just by having `SignData` take a [`crypto.SignerOpts`](https://golang.org/pkg/crypto/#SignerOpts) and then have a way to produce a TPM2 signing scheme from a SignerOpts.

Reply:
Sure, depending on architecture I could see the full signing scheme (not just the hash algorithm, though now we are definitely just talking about RSA signatures) being decided by a different component than the component that creates/attests the key.

However, we're just talking about RSA signatures in practice, and I'm less inclined to push for a super-flexible scheme just for RSA here if it adds complexity. (Especially since only non-attestation keys can have null schemes, and there is already the option of using the existing Sign method + hashing the data yourself.)

Reply:
OK that makes sense. I'll add functionality in a followup CL to make the `Signer` workflow function correctly for AlgNull keys. I need to have https://github.com/google/go-tpm/pull/226 merged first though.

Reply:
After my followup CL, we will have:
 - `SignData` will work with restricted and unrestricted keys, but they'll have to have a non-Null signing scheme
 - `GetSigner` + `Sign` will work with any unrestricted key, but you have to hash the data yourself (as is standard in Go)